flax/linen/recurrent.py

```diff
@@ -499,3 +510,205 @@ def initialize_carry(rng, batch_dims, size, init_fn=zeros):
+  contiguous and padded elements must be to the right of the sequence. For example::
+
+  TODO: improve example
+
+    >>> # 3 sequences with max length 5
+    >>> segmentation_mask = jnp.array([
+    ...   [1, 1, 1, 0, 0], # length 3
+    ...   [1, 1, 0, 0, 0], # length 2
+    ...   [1, 1, 1, 1, 1], # length 5
+    ... ])
+
+  The output will elements corresponding to padding elements are NOT zeroed out. If ``return_carry``
+  is set to ``True`` the carry will be the state of the last valid element of each sequence.
+
+  RNN also accepts some of the arguments of :func:`flax.linen.scan`, by default they are set to
+  work with cells like :class:`LSTMCell` and :class:`GRUCell` but you can be overriden as needed.
+  Specifying the default values would look like this::
+
+    >>> lstm = nn.RNN(
+    ...   nn.LSTMCell(), cell_size=64,
+    ...   unroll=1, variable_axes={}, variable_broadcast='params',
+    ...   variable_carry=False, split_rngs={'params': False})
+
+  Attributes:
+    cell: an instance of :class:`RNNCellBase`.
+    cell_size: the size of the cell as requested by :meth:`RNNCellBase.initialize_carry`,
+      it can be an integer or a tuple of integers.
+    time_major: if ``time_major=False`` (default) it will expect inputs with shape
+      ``(*batch, time, *features)``, else it will expect inputs with shape ``(time, *batch, *features)``.
+    return_carry: if ``return_carry=False`` (default) only the output sequence is returned,
+      else it will return a tuple of the final carry and the output sequence.
+    unroll: how many scan iterations to unroll within a single iteration of a loop,
+      defaults to 1. This argument will be passed to `nn.scan`.
+    variable_axes: the variable collections that are scanned over, this argument will
+      be passed to `nn.scan`.
+    variable_broadcast: specifies the broadcasted variable collections, this argument
+      will be passed to `nn.scan`.
+    variable_carry: specifies the variable collections that are carried through
+      the loop, this argument will be passed to `nn.scan`.
+    split_rngs: a mapping from PRNGSequenceFilter to bool specifying whether to
+      a PRNG collection should be split such that its values are different at each
+      step, or replicated such that its values remain the same at each step. This
+      argument will be passed to `nn.scan`.
+  """
+  cell: RNNCellBase
+  cell_size: Union[int, Tuple[int, ...]]
+  time_major: bool = False
+  return_carry: bool = False
+  unroll: int = 1
+  variable_axes: Mapping[lift.CollectionFilter,lift.InOutScanAxis] = FrozenDict()
+  variable_broadcast: lift.CollectionFilter = 'params'
+  variable_carry: lift.CollectionFilter = False
+  split_rngs: Mapping[lift.PRNGSequenceFilter, bool] = FrozenDict({'params': False})
+
+  def __call__(
+    self,
+    inputs: jax.Array,
+    *,
+    initial_carry: Optional[Carry] = None,
+    init_key: Optional[random.KeyArray] = None,
```

Comment:
Makes sense in this context, currently all of our cells initialize their carries using `zeros` so they effectively ignore it. Requiring it would be slightly annoying.